File: azmonitor/jobs/params.py
"""What a report request may ask for, checked the same way wherever it arrives.

The web application builds a request from a form; the worker receives only a job id and reads the
request back from the database. Both run it through these rules, so a value the browser could
never have sent — a sector that does not exist, a period in an unexpected shape, a key nobody
defined — is refused before it reaches the engine, however it got into the row.

The rules are deliberately few. A request names a report, a period (or "latest"), a sector for a
sector review, a publication for a brief, and whether to check the sources first. Nothing in it is a
path, a URL, a command, a repository or a branch.
"""
from __future__ import annotations

import datetime as dt
import re
from typing import Any

from .. import config

REPORT_TYPES: dict[str, dict[str, Any]] = {
    "monthly": {"label": "Monthly Monitor", "period": "month"},
    "weekly": {"label": "Weekly Digest", "period": "week"},
    "sector": {"label": "Sector Review", "period": "month", "sector": True},
    "mpr_brief": {"label": "MPR Brief", "publication_type": "monetary_policy_review"},
    "fsr_brief": {"label": "FSR Brief", "publication_type": "financial_stability_report"},
    "decision_update": {"label": "Policy Decision Update", "publication_type": "policy_decision"},
}
REFRESH = ("latest_data", "check_sources")
ALLOWED_KEYS = {"period", "sector", "publication_id", "refresh"}

_MONTH = re.compile(r"^\d{4}-(0[1-9]|1[0-2])$")
_DAY = re.compile(r"^\d{4}-\d{2}-\d{2}$")
# publication ids are "<type>:<edition>" as the engine assigns them, e.g. policy_decision:2026-07-24
_PUBLICATION = re.compile(r"^[a-z_]{3,40}:[A-Za-z0-9_.\-]{1,60}$")


class InvalidRequest(ValueError):
    def __init__(self, code: str, message: str):
        super().__init__(message)
        self.code = code
        self.message = message


def sectors() -> list[str]:
    return list((config.reports_config().get("sector") or {}).get("sectors") or {})


def normalise(report_type: str, raw: dict[str, Any] | None) -> dict[str, Any]:
    """The canonical parameters for a request, or InvalidRequest saying what is wrong.

    Canonical means the same request always produces the same dictionary, so two people asking
    for the same thing produce the same equivalence key and share one job.
    """
    if report_type not in REPORT_TYPES:
        raise InvalidRequest("unknown_report", f"There is no report type called {report_type!r}.")
    raw = dict(raw or {})
    extra = sorted(set(raw) - ALLOWED_KEYS)
    if extra:
        raise InvalidRequest("unknown_parameter", f"Unexpected parameter(s): {', '.join(extra)}.")
    spec = REPORT_TYPES[report_type]
    out: dict[str, Any] = {}

    refresh = raw.get("refresh") or "latest_data"
    if refresh not in REFRESH:
        raise InvalidRequest("invalid_refresh", "Choose either the latest collected data or a source check first.")
    out["refresh"] = refresh

    if spec.get("publication_type"):
        if raw.get("period") not in (None, "", "latest"):
            raise InvalidRequest("unsupported_period",
                                 f"A {spec['label']} is about one publication; choose the publication, not a period.")
        pub = raw.get("publication_id") or "latest"
        if pub != "latest":
            if not _PUBLICATION.match(str(pub)) or not str(pub).startswith(spec["publication_type"] + ":"):
                raise InvalidRequest("invalid_publication",
                                     f"{pub!r} is not a {spec['label']} source publication.")
        out["publication_id"] = pub
        if raw.get("sector"):
            raise InvalidRequest("unknown_parameter", "A sector applies only to a Sector Review.")
        return out

    period = raw.get("period") or "latest"
    if period != "latest":
        if spec["period"] == "month" and not _MONTH.match(str(period)):
            raise InvalidRequest("invalid_period", "A monthly period is written YYYY-MM, for example 2026-07.")
        if spec["period"] == "week":
            if not _DAY.match(str(period)):
                raise InvalidRequest("invalid_period", "A week is named by its Monday, written YYYY-MM-DD.")
            try:
                monday = dt.date.fromisoformat(str(period))
            except ValueError:
                raise InvalidRequest("invalid_period", f"{period} is not a date.") from None
            if monday.weekday() != 0:
                raise InvalidRequest("invalid_period",
                                     f"{period} is a {monday.strftime('%A')}; a week is named by its Monday.")
    out["period"] = period

    if spec.get("sector"):
        sector = raw.get("sector")
        if not sector:
            raise InvalidRequest("missing_sector", "Choose the sector to review.")
        if sector not in sectors():
            raise InvalidRequest("unknown_sector",
                                 f"There is no configured sector {sector!r}; choose one of {', '.join(sectors())}.")
        out["sector"] = sector
    elif raw.get("sector"):
        raise InvalidRequest("unknown_parameter", "A sector applies only to a Sector Review.")
    if raw.get("publication_id"):
        raise InvalidRequest("unknown_parameter", "A source publication applies only to the briefs.")
    return out


def scope_of(report_type: str, params: dict[str, Any]) -> str:
    """The thing an edition is *of*: a sector, a publication, a week or the latest month."""
    return params.get("sector") or params.get("publication_id") or params.get("period") or "latest"


def describe(report_type: str, params: dict[str, Any]) -> str:
    spec = REPORT_TYPES.get(report_type, {"label": report_type})
    bits = [spec["label"]]
    if params.get("sector"):
        bits.append(params["sector"])
    if params.get("publication_id") and params["publication_id"] != "latest":
        bits.append(params["publication_id"].split(":", 1)[-1])
    if params.get("period") and params["period"] != "latest":
        bits.append(params["period"])
    return " - ".join(bits)

File: web/lib/params.ts
/**
 * What a report request may ask for — the TypeScript twin of azmonitor/jobs/params.py.
 *
 * Both sides apply the same rules to the same cases (tests/fixtures/request-cases.json is read by
 * the Python and the Node tests alike), so the browser form, this API and the worker cannot
 * disagree about what a valid request is. The worker re-validates what it reads from the job row,
 * so a request that somehow bypassed this file is still refused before the engine runs.
 */
export const REPORT_TYPES = {
  monthly: { label: "Monthly Monitor", period: "month" },
  weekly: { label: "Weekly Digest", period: "week" },
  sector: { label: "Sector Review", period: "month", sector: true },
  mpr_brief: { label: "MPR Brief", publicationType: "monetary_policy_review" },
  fsr_brief: { label: "FSR Brief", publicationType: "financial_stability_report" },
  decision_update: { label: "Policy Decision Update", publicationType: "policy_decision" },
} as const;

export type ReportType = keyof typeof REPORT_TYPES;

/** The configured sectors (config/reports.yaml). tests/test_request_params.py checks they agree. */
export const SECTORS = ["agriculture", "construction", "trade", "transport", "industry"] as const;
export const REFRESH = ["latest_data", "check_sources"] as const;
const ALLOWED_KEYS = new Set(["period", "sector", "publication_id", "refresh"]);

const MONTH = /^\d{4}-(0[1-9]|1[0-2])$/;
const DAY = /^\d{4}-\d{2}-\d{2}$/;
const PUBLICATION = /^[a-z_]{3,40}:[A-Za-z0-9_.-]{1,60}$/;

export interface Params {
  refresh: (typeof REFRESH)[number];
  period?: string;
  sector?: string;
  publication_id?: string;
}

export class InvalidRequest extends Error {
  code: string;

  constructor(code: string, message: string) {
    super(message);
    this.code = code;
  }
}

export function isReportType(value: unknown): value is ReportType {
  return typeof value === "string" && Object.prototype.hasOwnProperty.call(REPORT_TYPES, value);
}

/** The canonical parameters, or InvalidRequest. `sectors` is the configured list. */
export function normalise(reportType: string, raw: Record<string, unknown> | null | undefined,
  sectors: readonly string[]): Params {
  if (!isReportType(reportType)) {
    throw new InvalidRequest("unknown_report", `There is no report type called '${reportType}'.`);
  }
  const input = { ...(raw ?? {}) } as Record<string, unknown>;
  for (const k of Object.keys(input)) {
    if (input[k] === undefined || input[k] === null || input[k] === "") delete input[k];
  }
  const extra = Object.keys(input).filter((k) => !ALLOWED_KEYS.has(k)).sort();
  if (extra.length) {
    throw new InvalidRequest("unknown_parameter", `Unexpected parameter(s): ${extra.join(", ")}.`);
  }
  const spec = REPORT_TYPES[reportType] as { label: string; period?: string; sector?: boolean;
    publicationType?: string };
  const refresh = (input.refresh ?? "latest_data") as string;
  if (!(REFRESH as readonly string[]).includes(refresh)) {
    throw new InvalidRequest("invalid_refresh",
      "Choose either the latest collected data or a source check first.");
  }
  const out: Params = { refresh: refresh as Params["refresh"] };

  if (spec.publicationType) {
    if (input.period !== undefined && input.period !== "latest") {
      throw new InvalidRequest("unsupported_period",
        `A ${spec.label} is about one publication; choose the publication, not a period.`);
    }
    const pub = String(input.publication_id ?? "latest");
    if (pub !== "latest" && (!PUBLICATION.test(pub) || !pub.startsWith(`${spec.publicationType}:`))) {
      throw new InvalidRequest("invalid_publication", `'${pub}' is not a ${spec.label} source publication.`);
    }
    out.publication_id = pub;
    if (input.sector !== undefined) {
      throw new InvalidRequest("unknown_parameter", "A sector applies only to a Sector Review.");
    }
    return out;
  }

  const period = String(input.period ?? "latest");
  if (period !== "latest") {
    if (spec.period === "month" && !MONTH.test(period)) {
      throw new InvalidRequest("invalid_period", "A monthly period is written YYYY-MM, for example 2026-07.");
    }
    if (spec.period === "week") {
      if (!DAY.test(period)) {
        throw new InvalidRequest("invalid_period", "A week is named by its Monday, written YYYY-MM-DD.");
      }
      const d = new Date(`${period}T00:00:00Z`);
      if (Number.isNaN(d.getTime()) || d.toISOString().slice(0, 10) !== period) {
        throw new InvalidRequest("invalid_period", `${period} is not a date.`);
      }
      if (d.getUTCDay() !== 1) {
        const name = d.toLocaleDateString("en-GB", { weekday: "long", timeZone: "UTC" });
        throw new InvalidRequest("invalid_period", `${period} is a ${name}; a week is named by its Monday.`);
      }
    }
  }
  out.period = period;

  if (spec.sector) {
    const sector = input.sector;
    if (!sector) throw new InvalidRequest("missing_sector", "Choose the sector to review.");
    if (typeof sector !== "string" || !sectors.includes(sector)) {
      throw new InvalidRequest("unknown_sector",
        `There is no configured sector '${String(sector)}'; choose one of ${sectors.join(", ")}.`);
    }
    out.sector = sector;
  } else if (input.sector !== undefined) {
    throw new InvalidRequest("unknown_parameter", "A sector applies only to a Sector Review.");
  }
  if (input.publication_id !== undefined) {
    throw new InvalidRequest("unknown_parameter", "A source publication applies only to the briefs.");
  }
  return out;
}

export function scopeOf(params: Params): string {
  return params.sector ?? params.publication_id ?? params.period ?? "latest";
}

export function describe(reportType: string, params: Partial<Params>): string {
  const spec = isReportType(reportType) ? REPORT_TYPES[reportType] : { label: reportType };
  const bits: string[] = [spec.label];
  if (params.sector) bits.push(params.sector);
  if (params.publication_id && params.publication_id !== "latest") bits.push(params.publication_id.split(":").slice(1).join(":"));
  if (params.period && params.period !== "latest") bits.push(params.period);
  return bits.join(" - ");
}
